Added ability to change the working directory via a command-line flag

dbonev opened the pull request:
Using the --cwd PATH option from the CLI, it is now possible to change the working directory to a path different from that of the compose file.

Implements #2586

Signed-off-by: Dimitar Bonev dimitar.bonev@gmail.com

dbonev commented:
Hmm... all tests look fine on my machine. It's not clear to me why the Jenkins build failed (also the errors look somewhat strange and unrelated).

dnephin commented:
I've restarted the build, I think the failures were unrelated.

LGTM when tests are green

aanand commented on compose/config/config.py:
```diff
@@ -133,10 +133,10 @@ def with_abs_paths(cls, working_dir, filename, name, config):
     if filenames == ['-']:
         return ConfigDetails(
-            os.getcwd(),
+            os.getcwd() if not override_dir else override_dir,
```
Could this be shortened to `override_dir or os.getcwd()`?

aanand commented on compose/config/config.py:
```diff
@@ -146,7 +146,7 @@ def find(base_dir, filenames):
     log.debug("Using configuration files: {}".format(",".join(filenames)))
     return ConfigDetails(
-        os.path.dirname(filenames[0]),
+        os.path.dirname(filenames[0]) if not override_dir else override_dir,
```
Similarly, could this be shortened to `override_dir or os.path.dirname(filenames[0])`?

aanand commented:
Should we have an acceptance test for this?

dnephin commented:
Ya, I guess the test doesn't cover all the code changes

dbonev commented:
@aanand, @dnephin Many thanks for the feedback.
I changed the code as requested and added acceptance tests. Unfortunately, the Jenkins build failed with an (I think) unrelated timeout error in a totally unrelated test. :disappointed: 

Again, it's all green on my machine, but please let me know if I'm missing something or something doesn't look right.

dnephin commented:
I think it's unrelated, I restarted the build

dbonev commented:
@dnephin Thanks

dnephin commented:
LGTM

dbonev commented:
Synced to the ```HEAD```. The test failure seems unrelated.

simonvanderveldt commented:
@dbonev What does `cwd` stand for? Compose working directory?
Isn't this more of a compose project thing and thus more analogous to `project_name` should be named `project_directory`?
Especially since by default the project name is the directory name containing the `docker-compose.yml` file (or the one where `docker-compose` is run from?).

Asking as I think it will probably be pretty confusing having different types of `working` directories.

Also, what's the exact use case for this? Does it include building and volume mounting `./<filename>` as well? And override/additional `docker-compose.*.yml` files?

dnephin commented:
Considering that I already made that mistake, I think I agree. Calling it project-directory might be good.

I think a good use-case is when you use stdin instead of a file path, you can set the project directory to something other than the current working directory.

dbonev commented:
OK, I'll resubmit as soon as I can. I'll change ```cwd``` to ```project-directory```.

dbonev commented:
Done. Sorry for the delay. Kinda overloaded lately.

dnephin commented on compose/cli/main.py:
```diff
@@ -144,6 +144,8 @@ class TopLevelCommand(DocoptCommand):
       -f, --file FILE           Specify an alternate compose file (default: docker-compose.yml)
       -p, --project-name NAME   Specify an alternate project name (default: directory name)
       --verbose                 Show more output
+      --project-directory PATH                Specify an alternate working directory
```
looks like the formatting here is off

dnephin commented:
Just one small issue with formatting, otherwise LGTM

aanand commented:
Yep - LGTM after the formatting fix

simonvanderveldt commented:
I think it would be a good idea to add a test to make sure that a volume mounted file from the project-dir works.

dnephin commented on compose/cli/main.py:
```diff
@@ -144,6 +144,8 @@ class TopLevelCommand(DocoptCommand):
       -p, --project-name NAME   Specify an alternate project name (default: directory name)
       --verbose                 Show more output
+      --project-directory PATH  Specify an alternate working directory
+                                (default: the path of the compose file)
```
hmm, I think this should say "the directory of the compose file" or "the dirname of the compose file".

dnephin commented on compose/config/config.py:
```diff
@@ -205,10 +205,10 @@ def with_abs_paths(cls, working_dir, filename, name, config):
     if filenames == ['-']:
         return ConfigDetails(
-            os.getcwd(),
+            override_dir or os.getcwd(),
```
I just noticed something which I missed earlier.

We already have `base_dir` here. I think setting `project-directory` should be equivalent to setting `base_dir`. I think we can probably re-use it. Instead of using `os.getcwd()`, this can be `os.abspath(base_dir)`, so when we pass in the default of `.` it will be expanded to the same path.

Sorry I didn't notice this earlier.

bfirsh commented:
Thanks for the contribution, @dbonev! Would you mind addressing the comments and rebasing, then we can get this in. 💃

dbonev commented:
@bfirsh  Yeah, I'll look at it asap. Sorry for the long delay (totally my bad, been quite busy lately).

andrerom commented:
Side: Is this only to be able to use different `.env` files?

If so a perhaps more direct fix would be to be able to set path to `.env` file *(including different name so you can have `.env_dev`)* either via argument and/or environment variable like partly proposed in description of #3399.

shin- commented:
Rebased and merged in #4589 - Thank you!
